include/bout/solver.hxx

```diff
@@ -350,6 +350,13 @@ public:
     pargv = &v;
   }
 
+  /// Generate a random UUID (version 4) and broadcast it to all processors
+  std::string createRunID() const;
```

Comment:
Maybe we want to keep `createRunID()` private, to ensure it only gets called once?

I wondered if we could just put a check in `createRunID()` to see if it had already been called, and if it has then just return `run_id`, but `getRunRestartFrom()` always needs to be called after the restart files have been loaded (and if `run_id` is set before restart files are loaded then it would be over-written, which we don't want). Anywhere that wants to save run_id should ideally also save run_restart_from, so it makes sense to restrict their use to after restart files are loaded, which is pretty much the same as after `createRunID()` is called. So I'd suggest making `createRunID()` private, and making it an error to call `getRunID()` or `getRunRestartFrom()` before `createRunID()`. We could add an extra flag to record when `createRunID()` is called, but maybe it'd be good enough to just check whether `run_id` still has its default value - that check would be wrong between the restarts being loaded and `createRunID()` being called, but probably nothing could happen in that time? Or maybe it's better to have a specific flag to be safe and future proof?

Reply:
Good points. Maybe just have `getRunID` and `getRunRestartFrom` throw if `run_id` is still its default value?

The guts of `createRunID` may be more generally useful if we want UUIDs for other things (not that I can think of any off the top of my head), but yes, should be private for now.

Reply:
> Maybe just have getRunID and getRunRestartFrom throw if run_id is still its default value?

sounds good to me.